### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by stas7t**

### CODE REVIEW

File: lib/codebreaker.rb [hunk 1]
```diff
@@ -1,5 +1,53 @@
...
+  def play
+    rules = File.open('./data/rules.txt').read
+    rules.each_line do |line|
+      puts line
+    end
+
+    game = Game.new
+    game.start
+
+    user_input = proc do
```
Comment (yamnich) [inline lib/codebreaker.rb]: too long method, could we separate it somehow?

File: lib/codebreaker/game.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      return false unless code.length == 4 && code.match(/[1-6]{4}/)
+      @attempts -= 1
+      @user_guess = code
+    end
+
+    def mark_result
+      return '++++' if @secret_code == @user_guess
+      result = '....'
+      @secret_code.chars.each_index do |i|
+        @user_guess.chars.each_index do |j|
```
Comment (yamnich) [inline lib/codebreaker/game.rb]: could we solve it without 2 each

File: spec/codebreaker/game_spec.rb [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+          .to change { game.instance_variable_get(:@attempts) }.by(-1)
+      end
+    end
+
+    context '#mark_result' do
+      before do
+        game.instance_variable_set(:@secret_code, '1234')
+      end
+
+      it 'returns `++++` if secret code is guessed' do
```
Comment (yamnich) [inline spec/codebreaker/game_spec.rb]: please add more cases

**Comment by stas7t** [comment]
1. Fix marking algorithm
2. Add 204 test cases for marking algorithm
3. Fix long method (play). 
4. Add UI class and tests.

File: lib/codebreaker/game.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    def result
+      return 'win' if win?
+      'lose' if lose?
+    end
+
+    def reveal_code
+      @secret_code if win? || lose?
+    end
+
+    def play_again?(answer)
```
Comment (yamnich) [inline lib/codebreaker/game.rb]: this is should not be in game

**Pull request merged**